packages/file-collections/src/node/TempFileStoreWorker.js

```diff
@@ -19,49 +19,66 @@ function forEachPromise(items, fn) {
 }
 
 export default class TempFileStoreWorker extends EventEmitter {
-  constructor({ fileCollections = [] } = {}) {
+  constructor({ fileCollections = [], onNewFileRecord } = {}) {
     super();
 
     this.fileCollections = fileCollections;
     this.observeHandles = [];
+    this.onNewFileRecord = onNewFileRecord;
+  }
+
+  pushObservedDocument(doc, collection, stores, tempStore) {
+    const { onNewFileRecord } = this;
+    if (onNewFileRecord) {
+      onNewFileRecord(doc, collection);
+    } else {
+      debug("No onNewFileRecord function passed to TempFileStoreWorker, using internal function");
+      this.handleTempStoreIdAdded({ collection, doc, stores, tempStore })
+        .catch((error) => {
+          console.error(error); // eslint-disable-line no-console
+        });
+    }
   }
 
   start() {
     this.fileCollections.forEach((collection) => {
       const { stores, tempStore } = collection.options;
 
-      // Support for storing to multiple stores from a tempStore
-      this.observeHandles.push(collection.mongoCollection.find({
-        "original.tempStoreId": { $ne: null }
-      }).observe({
-        added: (doc) => {
-          this._handleTempStoreIdAdded({ collection, doc, stores, tempStore })
-            .catch((error) => {
-              console.error(error); // eslint-disable-line no-console
-            });
-        },
-        removed(doc) {
-          const { tempStoreId } = doc.original || {};
-
-          debug(`TempFileStoreWorker: Removing ${tempStoreId} from TempFileStore`);
-
-          // Delete the file from the temp store if it exists
-          tempStore.deleteIfExists(tempStoreId).catch((error) => {
-            console.error(error); // eslint-disable-line no-console
-          });
+      const handle = collection.collection.watch([{
+        $match: {
+          "operationType": "insert",
+          "fullDocument.original.tempStoreId": { $ne: null }
```

Comment:
As far as the "removed" watcher, I think you aren't quite understanding how that works.

The "removed" function is called when a document disappears from the _result set_, and not from the collection. In other words, it's essentially continually re-evaluating the find query `{ "original.tempStoreId": { $ne: null } }`, and if a document was found last time and is not found this time, it's "removed" from the results.

That means that a document will only be passed to the "removed" function here if `tempStoreId` becomes `null` or undefined for a document.

Ref: https://docs.meteor.com/api/collections.html#Mongo-Cursor-observe

If you look near the end of `handleTempStoreIdAdded`, it is unsetting `tempStoreId`, which should be triggering the "removed" handler to delete the temp files.

We do need to delete the temporary files after we're done with them, but I don't remember why this was being done by observing the cursor. Since this is all in the same file, I suggest that you move the `tempStore.deleteIfExists` block down to the end of `handleTempStoreIdAdded`, just after the `update` that unsets the ID. This will keep the flow essentially as it was.

Reply:
Thanks for the clarification - Will put it there, didn't really know where to put it as I didn't understand its purpose.